File: core/src/test/java/com/oneandone/typedrest/ActionEndpointTest.java
package com.oneandone.typedrest;

import static com.github.tomakehurst.wiremock.client.WireMock.*;
import static org.apache.http.HttpHeaders.*;
import static org.apache.http.HttpStatus.*;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import org.junit.*;

public class ActionEndpointTest extends AbstractEndpointTest {

    private ActionEndpoint endpoint;

    @Before
    @Override
    public void before() {
        super.before();
        endpoint = new ActionEndpointImpl(entryEndpoint, "endpoint");
    }

    @Test
    public void testProbe() throws Exception {
        stubFor(options(urlEqualTo("/endpoint"))
                .willReturn(aResponse()
                        .withStatus(SC_OK)
                        .withHeader(ALLOW, "POST")));

        endpoint.probe();

        assertThat(endpoint.isTriggerAllowed().get(), is(true));
    }

    @Test
    @Ignore("Works in isolation but fails when executed as part of test suite")
    public void testTrigger() throws Exception {
        stubFor(post(urlEqualTo("/endpoint"))
                .willReturn(aResponse().withStatus(SC_OK)));

        endpoint.trigger();
    }
}

File: typedrest/src/test/kotlin/net/typedrest/endpoints/rpc/ActionEndpointTest.kt
package net.typedrest.endpoints.rpc

import net.typedrest.endpoints.AbstractEndpointTest
import net.typedrest.http.HttpMethod
import net.typedrest.tests.assertRequest
import okhttp3.mockwebserver.MockResponse
import kotlin.test.*

class ActionEndpointTest : AbstractEndpointTest() {
    private val endpoint = ActionEndpointImpl(entryEndpoint, "endpoint")

    @Test
    fun testProbe() {
        server.enqueue(MockResponse().setHeader("Allow", "POST"))

        endpoint.probe()

        server.assertRequest(HttpMethod.OPTIONS)
        assertEquals(true, endpoint.isInvokeAllowed)
    }

    @Test
    fun testInvoke() {
        server.enqueue(MockResponse())

        endpoint.invoke()
        server.assertRequest(HttpMethod.POST)
    }
}
